### PR TITLE
bug(cli): add support for repos that are submodules

### PR DESCRIPTION
**Pull request by maxakuru**
- Add `getGitDir()` function to GitUtils to find gitdir of a repo directory
- Use `getGitDir()` in UpCommand
- Add `gitdir` argument to all GitUtils methods that use isomorphic-git APIs that support gitdir option

### TODO
- [ ] tests for added branches

## Related Issues
#1794

### CODE REVIEW

File: src/git-utils.js [hunk 1]
```diff
@@ -30,14 +31,17 @@ class GitUtils {
    *
    * @param {string} dir working tree directory path of the git repo
    * @param {string} [homedir] optional users home directory
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:34]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/git-utils.js [hunk 2]
```diff
@@ -105,9 +109,10 @@ class GitUtils {
    * @param {string} dir working tree directory path of the git repo
    * @param {string} filepath file to check
    * @param {string} [homedir] optional users home directory
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:112]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/git-utils.js [hunk 3]
```diff
@@ -140,25 +145,35 @@ class GitUtils {
    * will be returned instead.
    *
    * @param {string} dir working tree directory path of the git repo
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:148]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/git-utils.js [hunk 4]
```diff
@@ -174,11 +189,12 @@ class GitUtils {
    * current branch or tag.
    *
    * @param {string} dir working tree directory path of the git repo
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:192]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/git-utils.js [hunk 5]
```diff
@@ -188,10 +204,11 @@ class GitUtils {
    * will be returned instead.
    *
    * @param {string} dir working tree directory path of the git repo
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:207]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/git-utils.js [hunk 6]
```diff
@@ -200,11 +217,12 @@ class GitUtils {
    * Returns the `origin` remote url or `''` if none is defined.
    *
    * @param {string} dir working tree directory path of the git repo
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:220]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/git-utils.js [hunk 7]
```diff
@@ -216,41 +234,50 @@ class GitUtils {
    * Same as #getOrigin() but returns a `GitUrl` instance instead of a string.
    *
    * @param {string} dir working tree directory path of the git repo
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:237]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/git-utils.js [hunk 8]
```diff
@@ -216,41 +234,50 @@ class GitUtils {
...
+  static async getOriginURL(dir, gitdir) {
+    const origin = await GitUtils.getOrigin(dir, gitdir);
     return origin ? new GitUrl(origin) : null;
   }
 
   /**
    * Returns the sha of the current (i.e. `HEAD`) commit.
    *
    * @param {string} dir working tree directory path of the git repo
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:249]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/git-utils.js [hunk 9]
```diff
@@ -216,41 +234,50 @@ class GitUtils {
...
+      fs, dir, ref: 'HEAD', gitdir,
+    });
   }
 
   /**
    * Returns the commit oid of the curent commit referenced by `ref`
    *
    * @param {string} dir git repo path
    * @param {string} ref reference (branch, tag or commit sha)
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:263]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/git-utils.js [hunk 10]
```diff
@@ -263,16 +290,74 @@ class GitUtils {
    * @param {string} dir git repo path
    * @param {string} ref reference (branch, tag or commit sha)
    * @param {string} filePath relative path to file
+   * @param {string} [gitdir] the git directory path
```
Comment (tripodsan) [inline src/git-utils.js:293]:
```suggestion
   * @param {string} [gitdir] optional git directory path
```

File: src/up.cmd.js [hunk 11]
```diff
@@ -179,10 +179,11 @@ class UpCommand extends BuildCommand {
 
   async setup() {
     await super.init();
-    // check for git repository
-    if (!await fse.pathExists(path.join(this.directory, '.git'))) {
-      throw Error('hlx up needs local git repository.');
+    if (!this._gitHead) {
+      this._gitDir = await GitUtils.getGitDir(this.directory);
+      this._gitHead = path.join(this._gitDir, 'HEAD');
```
Comment (tripodsan) [inline src/up.cmd.js:184]: why not `path.resolve()` ?

**Review by tripodsan — CHANGES_REQUESTED** [review]
would be good to add an integration test that runs hlx up from submodule directory.

**Comment by trieloff** [comment]
Only nitpick from my side: use `fix(cli):` instead of `bug(cli):` in the commit message, as only `fix`, `feat` and `BREAKING  CHANGE` increase the version counter and lead to new releases.

File: src/up.cmd.js (hunk 11 above)
Comment (maxakuru) [reply src/up.cmd.js:184]: no real reason, `this.directory` is already resolved and `getGitDir()` returns a resolved path - do you always prefer `path.resolve()`?

**Comment by tripodsan** [comment]
@maxakuru why did you close this PR ?

**Comment by maxakuru** [comment]
@tripodsan seemed like the consensus on #1794 was that submodule support wasn't wanted so closed this for #1796

**Pull request closed without merging**